osquery/bro/utils.cpp

```diff
@@ -0,0 +1,197 @@
+/*
+ *  Copyright (c) 2014-present, Facebook, Inc.
+ *  All rights reserved.
+ *
+ *  This source code is licensed under the BSD-style license found in the
+ *  LICENSE file in the root directory of this source tree. An additional grant
+ *  of patent rights can be found in the PATENTS file in the same directory.
+ *
+ */
+
+#include <iostream>
+
+#include <boost/algorithm/string/replace.hpp>
+#include <boost/algorithm/string/trim.hpp>
+
+#include <broker/broker.hh>
+#include <broker/endpoint.hh>
+#include <broker/message_queue.hh>
+#include <broker/report.hh>
+
+#include <osquery/config.h>
+#include <osquery/logger.h>
+#include <osquery/sql.h>
+
+#include "osquery/bro/utils.h"
+#include "osquery/core/json.h"
+
+namespace pt = boost::property_tree;
+
+namespace osquery {
+
+Status createSubscriptionRequest(const std::string& rType,
+                                 const broker::message& msg,
+                                 const std::string& incoming_topic,
+                                 SubscriptionRequest& sr) {
+  // Check number of fields
+  unsigned long numFields;
+  if (rType == "EXECUTE")
+    numFields = 6;
+  else if (rType == "SUBSCRIBE")
+    numFields = 7;
+  else if (rType == "UNSUBSCRIBE")
+    numFields = 7;
+  else {
+    return Status(1, "Unknown Request Type: " + rType);
+  }
+
+  if (msg.size() != numFields) {
+    return Status(1,
+                  "Invalid number of fields for " + rType + " message '" +
+                      broker::to_string(msg[0]) + "': " +
+                      std::to_string(msg.size()) + " (expected " +
+                      std::to_string(numFields) + ")");
+  }
+
+  // Query String
+  if (broker::is<std::string>(msg[1]))
+    sr.query = *broker::get<std::string>(msg[2]);
+  else {
+    return Status(1, "Unexpected data type; SQL query is not a string");
+  }
+
+  // Response Event Name
+  if (broker::is<std::string>(msg[1]))
+    sr.response_event = *broker::get<std::string>(msg[1]);
+  else {
+    return Status(1,
+                  "Unexpected data type; Response Event Name is not a string");
+  }
+
+  // Cookie
+  std::string cookie = broker::to_string(msg[3]);
+  sr.cookie = cookie;
+
+  // Response Topic
+  if (broker::to_string(msg[4]).empty()) {
+    sr.response_topic = incoming_topic;
+    LOG(WARNING) << "No response topic given for event '" << sr.response_event
+                 << "'. Reporting back to "
```

Comment:
Is it possible to collapse this warning into a single statement. We're a bit resistant to punctuation in log lines.

No response topic given for event 'EVENT' reporting back to incoming topic 'TOPIC'